source/shared/autofill.js

```diff
@@ -1,74 +1,104 @@
 import { NativeModules, Platform } from "react-native";
 import { EntryFinder } from "../library/buttercupCore";
+import { getEntryPathString } from "./entry";
 
 const { AutoFillBridge } = NativeModules;
 
-export function addSourceToAutoFill(sourceID, archive) {
+// Note: AutoFillBridge is not available in jest, and cannot be mocked without mocking
+//  every single NativeModule from node_modules as well (not something I have time to do)
+//  Instead, just check for both the existence of AutoFillBridge and the value of DEVICE_SUPPORTS_AUTOFILL
+//  - @se1exin 17/2/19
+export const autoFillAvailable = AutoFillBridge && !!AutoFillBridge.DEVICE_SUPPORTS_AUTOFILL;
+
+export function getAutoFillSystemStatus() {
+    return new Promise((resolve, reject) => {
+        if (autoFillAvailable) {
+            return AutoFillBridge.getAutoFillSystemStatus().then(isAutoFillEnabled => {
+                resolve(isAutoFillEnabled);
+            });
+        }
+        resolve(false);
+    });
+}
+
+export function openAutoFillSystemSettings() {
     return new Promise((resolve, reject) => {
-        if (Platform.OS !== "ios" || !AutoFillBridge.DEVICE_SUPPORTS_AUTOFILL) {
-            // AutoFill is only implemented on iOS at the moment
-            resolve();
-            return;
+        if (autoFillAvailable) {
+            return AutoFillBridge.openAutoFillSystemSettings().then(() => {
+                resolve();
+            });
         }
+        resolve(false);
+    });
+}
 
-        // We need to flatten all the Archive Entries, then send them to the native module
-        const finder = new EntryFinder([archive]);
-        let entries = {}; // Entries will be keyed by ID
+export function autoFillEnabledForSource(sourceID) {
+    return new Promise((resolve, reject) => {
+        if (autoFillAvailable) {
+            return AutoFillBridge.getAutoFillEnabledSources().then(autoFillSources => {
```

Comment:
And this too.. Otherwise perhaps wrapping it in something like:

```javascript
return Promise.resolve()
	.then(() => {
		// ...
		return false;
	});
```

Reply:
Not sure about this one - the promise resolves on all paths, unless I'm missing something?